File: LeetCode/39-Combination-Sum(Medium)/Solution.java
class Solution {
    private List<List<Integer>> res = new ArrayList<>();
    
    private void combinationSumHelper(int[] candidates, List<Integer> list, int start, int remain) {
        if (remain < 0) return;
        
        if (remain == 0) {
            res.add(new ArrayList<Integer>(list));
            return;
        }
        
        for (int i = start; i < candidates.length; ++i) {
            list.add(candidates[i]);
            combinationSumHelper(candidates, list, i, remain - candidates[i]);
            list.remove(list.size() - 1);
        }
        
    }
    
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<Integer> list = new ArrayList<>();
        
        combinationSumHelper(candidates, list, 0, target);
        
        return res;
    }
}

File: LeetCode/39-Combination-Sum(Medium)/Solution.cpp
class Solution {
private:
    void dfs(vector<int>& candidates, vector<vector<int>>& res, vector<int>& list, int start, int target) {
        if (target < 0) return;
        if (target == 0) {
            res.push_back(list);
            return;
        }
        
        for (int  i = start; i < candidates.size(); ++i) {
            list.push_back(candidates[i]);
            dfs(candidates, res, list, i, target - candidates[i]);
            list.erase(list.end() - 1);
        }
    }
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> list;
        vector<vector<int>> res;
        
        dfs(candidates, res, list, 0, target);
        
        return res;
    }
};